### ISSUE TITLE
Support server-side-only LDAP

### ISSUE THREAD

**Issue by alkisg** [ltsp/ltsp#16]
Since the new LTSP uses PAM, LDAP is kind of already supported...
...if someone configures LDAP in the chroot/VM for authentication,
...and uses NFS3 for insecure /home,
...or NFS4/kerberos etc for secure /home.

But we also want to support the case where one configures LDAP on the LTSP server, and wants the clients to authenticate via SSH on the LTSP server, not via LDAP.

Currently for this to work, the sysadmin needs to generate /etc/ltsp/{passwd,group} files from LDAP, maybe from an `ltsp initrd` hook. It needs to become easier and more documented.

**Comment by MAccadia** [comment]
_I had a look to `pamnotes.md` but I don't fully understand the authentication design. Having all users' logins in /etc/passwd on the server seems unpractical for thousands of logins. Anyway, here are some ideas about this issue._

If we use the standard display managers and PAM, it makes sense to use a pam module to authenticate.

Maybe it's possible to turn the LTSP server into an LDAP proxy or broker and use pam_ldap / pam_sss to authenticate against the LTSP server. An other solution may be to make or use a pam module that use the ability to log into an SSH server as a way to authenticate.

**Comment by alkisg** [comment]
We need two things: authentication and home directory.

In big setups with thousands of users, the recommended way would be to configure pam_ldap in the chroot, and kerberos, and nfs4. Then pamltsp and sshfs home directories aren't needed or used at all.

In small setups that don't use ldap, pamltsp and sshfs already work fine.

The problem is if someone wants to use ldap with sshfs. No current pam module can do that. We would need to write one in C, and upload it to Debian and the other distributions, and make it a dependency of the ltsp client package.

Help wanted in either implementing or sponsoring that.

In the meantime, the current pamltsp implementation allows sysadmins to export ldap users to /etc/ltsp/passwd and /etc/ltsp/group, and pamltsp will pick them up from there.

Since pamltsp is using pam_exec, we cannot dynamically create users locally while authenticating them. That's why the user account must exist first.
We *could* create the user while authenticating, but then:

- the user would try to login once, and fail,
- pamltsp would then create that account,
- and the second login attempt would succeed.

To sum up:

1. We either need the users in passwd/group
2. Or ldap+kerberos+nfs4 needs to be configured in the chroot/image
3. Or we need to write ldapssh modules in C
4. Or "unknown" users will need to try to login twice
5. Or we'd need to write our own display manager and not rely on PAM, like LTSP5/LDM did (and that was the main thing we wanted to avoid in LTSP19)

If you're saying that "1" is impractical, even with a nice script that will export ldap users into passwd/group files, then "2" or "3" are probably necessary. For "3", someone will be needed to either implement or sponsor it.

**Comment by MAccadia** [comment]
Thank you for the clarifications. I think ldap+kerberos+nfs4 is indeed the best for big setups.

I said that users in `passwd/group` is unpractical but I don't really know how many users standard unix authentication can handle. Sometimes flat files are as good as databases.

**Comment by alkisg** [comment]
I think pam_unix (i.e. passwd/group) can easily handle thousands of users.

I just thought of another option though,

6. To configure ldap in the chroot/image to provide NSS services; that way all user accounts will "exist" in the client and pamltsp won't need to create them; it'll just need to do the sshfs mount.

**Comment by MAccadia** [comment]
If I understood correctly, the issue is not much "Support server-side-only LDAP" than "Support SSHFS homes with LDAP auth".

Why have pam_mount been ruled out ? Is it because ssh need a homedir to work ?

**Comment by alkisg** [comment]
For pam_mount to work with sshfs, something like this would be needed:
https://debian-administration.org/article/587/pam_mount_and_sshfs_with_password_authentication

But pamltsp already does that and more, so pam_mount wouldn't offer anything there.

Both pamltsp and pam_mount assume that the user exists on the client. This is the main problem.

Currently, ltsp solves it by getting passwd and group from the server, while pam_mount doesn't offer any solution.

To better analyze solution "6" above:
- The administrator would configure ldap nss services in the chroot, so that all users would "exist" in the clients. I.e. even before login, `getent passwd some-user` would return a proper entry.
- And pamltsp then would work fine, without relying on passwd/group, as it would only need to do the sshfs mount, as it already does.

So I think this issue is more about documenting all that; the pamltsp changes for solutions "2" or "6" would be minimal.

**Comment by alkisg** [comment]
Continuing from duplicate issue #26.

At this stage, someone that uses LDAP needs to give feedback and I'll try to help in order to get this working.
We could try to solve this by writing in this issue, but if you don't mind, it would be way faster if you came to IRC and we shared your screen for a while (over Epoptes/reverse VNC, no need for port forward etc).

The point of failure there is because LDAP modified /etc/pam.d/common-auth, and LTSP no longer knows how to install itself.

If you were to use NFS, you could just disable pamltsp completely by creating an empty /etc/ltsp/client/init/55-pam.sh.
But if you want to use (the default) SSHFS, you'll need pamltsp to mount /home/username.

That means that a few lines of code need to be changed, so that pamltsp installs itself properly in common-auth even if you have LDAP there, and so that it doesn't try to handle /etc/{passwd,group} at all. Help is needed there

**Comment by wforumw** [comment]
we are using NFS to share users home directory

**Comment by wforumw** [comment]
I can only test it on Monday.
I keep you informed.
Many thanks for the quick response.

**Comment by wforumw** [comment]
Authentication works now

**Comment by wforumw** [comment]
Authentication works now.
Thanks

**Comment by alkisg** [comment]
@wforumw, great!

Could you paste the whole files that contain "ldap" under /etc/pam.d, so that I configure pamltsp to do nothing in this case?

**Comment by Uumas** [comment]
Here's entries when using libpam-ldap(d):
```
uumas@ltsp19:~$ grep -r ldap /etc/pam.d
/etc/pam.d/common-session-noninteractive:session        [success=ok default=ignore]   pam_ldap.so minimum_uid=1000
/etc/pam.d/common-auth:auth     [success=1 default=ignore]   pam_ldap.so minimum_uid=1000 use_first_pass
/etc/pam.d/common-password:password     [success=1 default=ignore]    pam_ldap.so minimum_uid=1000 try_first_pass
/etc/pam.d/common-session:session       [success=ok default=ignore]   pam_ldap.so minimum_uid=1000
/etc/pam.d/common-account:account       [success=ok new_authtok_reqd=done ignore=ignore user_unknown=ignore authinfo_unavail=ignore default=bad]    pam_ldap.so minimum_uid=1000
```

Here's the entries when using sssd:
```
uumas@ltsp19:~$ grep -r sss /etc/pam.d
/etc/pam.d/common-auth:auth     [success=2 default=ignore]   pam_sss.so use_first_pass
/etc/pam.d/common-password:password     sufficient           pam_sss.so use_authtok
/etc/pam.d/common-session:session       optional             pam_sss.so
/etc/pam.d/common-account:account       [default=bad success=ok user_unknown=ignore]  pam_sss.so
```

**Comment by wforumw** [comment]
@alkisg, we are using sssd.
Uumas gave already the pam entries

**Comment by alkisg** [comment]
Great, thank you. At the moment I'll implement an easy way for pamltsp to be either manually or automatically disabled (easier than the empty 55-pam.sh file).

**Comment by wforumw** [comment]
Hi,
Very strange after update ltsp to 

> LTSP version: 19.09-1 201909270716

we have the same issue with LDAP again
> Could not configure PAM for SSH authentication!
> LTSP command failed: /usr/share/ltsp/client/login/pamltsp install

The file  /etc/ltsp/client/init/55-pam.sh file still exist and we run ltsp initrd

Did something changed?
Many Thanks!

**Comment by alkisg** [comment]
Yes sorry 55-pam.sh was renamed to 54-pam.sh, because the user accounts were needed before 55-display-manager.sh.

So rename /etc/ltsp/client/init/55-pam.sh to /etc/ltsp/client/init/54-pam.sh and run `ltsp initrd` again.

**Comment by wforumw** [comment]
Thanks!
Works again.

**Comment by feschoppe** [comment]
Hi,
at our school we are authenticating against a samba domain controller. With our present setup (LTSP5 with dual nic and isc-dhcp-server on internal nic, Ubuntu 18.04 Mate and Fat Clients) it works really fine. Authentication works and the shares are mounted automatically with cifs-utils. Unfortunately, I don’t get it working in a LTSP 19.08 test setup. I configured the server to authenticate against the dc and users are able to login into the server using ssh. But when I boot a client, the boot fails saying that there is some kind of trouble with pam authentication. After I created /etc/ltsp/client/init/54-pam.sh, ran ltsp initrd, the client boots properly, I get the login screen but authentication fails. So I can’t log in (domain users and local server users fail).
Any help is appreciated! By the way, thanks for your great work! We are using a LTSP setup for about a year now and it really works fine! We are expanding our infrastructure and I would love to use LTSP 19.08 in any new installation, but before we need to get authentication and mounting of samba shares working.

**Comment by alkisg** [comment]
Hi feschoppe,

are you using the chrootless mode, i.e. `ltsp image /`? If yes, you should be able to cancel 54-pam.sh like you did, then boot the client, then switch to vt1, and login there, with the domain user accounts, not the local ones.

At that point, graphical login might not work as /home/username isn't automounted with sshfs; 54-pam.sh was supposed to setup the sshfs mounts. That can be fixed with cifs or nfs later on.

But if vt login doesn't work, then it's up to you to fix it; the ltsp client at that point is just a clone of the server. Does the samba domain controller require a separate "join" procedure for each clone?

**Comment by feschoppe** [comment]
Hi,
thanks for your fast reply! I am using chrootless mode. In fact, I figured out that my client doesn’t have network access. I can’t ping the domain controler. There seems to be an issue with routing from the internal network to the outsinde network. So logging in using vt1 fails (no logon servers). This seems to be my problem.
Our samba dc does not require a seperate join procedure for each client.

**Comment by feschoppe** [comment]
I resolved my problem: I fixed my routing issue and I realized that I have to join each client separately to the domain controller. With LTSP5 this has not been necessary. I don’t know why I have to do it now, but it works and that is what is important. It even works with graphical login now.

**Comment by alkisg** [comment]
LTSP5 didn't have proper "PAM based" authentication. It had its own display manager called LDM, which got the username/password from the user, validated them via SSH to the server, and then it **created** a temporary user in the client /etc/passwd and `su -`'ed to that user.

But that meant that we had to maintain our own display manager, and maintain code on how to start it, how it would run the session, which XDG_* environment variables it would need to set, and now we'd have to rewrite it to support Wayland etc. We couldn't continue doing that with the limited development power that we have. This was the most significant reason to rewrite LTSP from scratch.

So the new implementation uses PAM and that makes it possible to use any display manager like LightDM or GDM or SDDM; the down side is that we cannot "fake" a local user anymore; the user account either needs to be in /etc/passwd before the authentication starts, or in LDAP.

(For potential contributors; if one develops a pamltsp **account** pam module in .c, then we can implement user accounts dynamically like LDAP does; but I don't have time to study PAM well enough to write such a module).

**Comment by feschoppe** [comment]
Thank you for this information!

I found a quick-and-dirty solution to automatically join the samba domain when starting a client: I created a systemd service unit which starts a short script to join the domain. Now we can authenticate and use the clients with our domain user accounts.

**Comment by alkisg** [comment]
It would be really nice if someone of you guys that use LDAP/AD/Samba etc wrote a wiki page about it.
LTSP is a small community currently, so it does need as much help as it can get. :)

https://github.com/ltsp/community/wiki

**Comment by feschoppe** [comment]
I wrote a little wiki page about my quick and dirty way of implementing samba domain authentication:

https://github.com/ltsp/community/wiki/LTSP-and-Samba-Windows-NT-Domain

**Comment by alkisg** [comment]
+10 karma points to @feschoppe for the great wiki page!

**Comment by ncswalters** [comment]
Thanks @feschoppe for the wiki page.  Has anyone found a more secure method for joining the clients to the domain on boot?  I know that you were looking for a better solution that doesn't require the domain admin password stored in the clear.  We are in the process of testing LTSP19 for the next school year, but that's a deal breaker.

**Comment by Uumas** [comment]
@ncswalters I've been thinking of just joining one client manually and copying that keytab and config to the ltsp image. That way only the keytab is public which isn't nearly as bad.

**Comment by MAccadia** [comment]
This is slightly off-topic (since this issue is about server-side auth) but it might be possible to authenticate against AD without joining the domain : the AD is used as a standard LDAP ([see this](https://medium.com/jerrynotes/linux-authentication-windows-ad-without-join-domain-7963c3fd44c5)) and authentication can be performed with Kerberos (sssd with `auth_provider = krb5`).

I would have tried that but I haven't admin rights on our AD.

**Comment by Uumas** [comment]
@MAccadia I'm currently doing this using nslcd (pure ldap) with a non-admin account. I assume it should work using krb5 similarly. What would you need an admin rights for?

**Comment by MAccadia** [comment]
> What would you need an admin rights for?

I don't have the admin rights to create the read-only non-admin account used for listing users (I guess the one you use in your nslcd config).

AFAIK krb5 is only used for authentication and can't be used as the user database (identification).

**Comment by Uumas** [comment]
@MAccadia You can use your personal account for testing (and of course only for testing!). And yes, krb5 is only for auth, ldap is used for getting account info. The same way as it would work on an enrolled host.

**Comment by ncswalters** [comment]
Howdy Gang.  Does anyone have a tutorial I can follow to get auth working between the clients and AD?  I’ve looked at the wiki page for the Samba / NT instructions but that still is sending the admin password in the clear.  The other option was to copy the keytab, which may work, but I’m unclear on.  Any help is greatly appreciated!

**Comment by alkisg** [comment]
I'll convert this into a discussion as it's superseded by #338.
